InteractionBase/src/vis/Menu.cpp

```diff
@@ -0,0 +1,245 @@
+		for (int row = 0; row < currentNodes_[col].size(); row++)
+			if (currentNodes_[col][row] == node)
+				return QPoint(col, row);
+	return QPoint(-1, -1);
+}
+
+QPoint Menu::correctCoordinates(QPoint point) const
+{
+	QPoint result = point;
+	if (point.x() < 0)
+		result.setX(currentNodes_.size() - 1);
+	if (point.x() >= currentNodes_.size())
+		result.setX(0);
+	if (point.y() < 0)
+		result.setY(currentNodes_[point.x()].size() - 1);
+	if (point.y() >= currentNodes_[result.x()].size())
+		result.setY(0);
+	return result;
+}
+
+void Menu::determineChildren()
+{
+	Super::determineChildren();
+	Item* defaultProxy{};
+	if (hasTextField())
+		setDefaultMoveCursorProxy(defaultProxy = textField_);
+	else
+		setDefaultMoveCursorProxy(defaultProxy = findVisualizationOf(currentNodes()[0][0]));
+	//TODO@cyril doing this in show doesn't work -> the grid does not have items yet
+	if (!focusedItem_)
+		selectItem(defaultProxy);
+}
+
+void Menu::updateGeometry(int availableWidth, int availableHeight)
+{
+	Super::updateGeometry(availableWidth, availableHeight);
+	setPos(mousePosition_.x() - widthInScene() / 2, mousePosition_.y() - heightInScene() / 2);
+}
+
+bool Menu::sceneEventFilter(QGraphicsItem* watched, QEvent* event)
+{
+	if (event->type() == QEvent::KeyPress)
+	{
+		auto keyEvent = static_cast<QKeyEvent*>(event);
+		if (keyEvent->key() == Qt::Key_Down || keyEvent->key() == Qt::Key_Up
+			|| keyEvent->key() == Qt::Key_Left || keyEvent->key() == Qt::Key_Right)
+		{
+			event->accept();
+			//If currently the name is selected, and down key is pressed -> select the item below
+			if (focusedItem() == textField() && keyEvent->key() == Qt::Key_Down)
+				selectItem(findVisualizationOf(currentNodes()[0][0]));
+			//If an item in the top row is selected, and up key is pressed -> select the name
+			else if (focusedItem() != textField() && indexOf(focusedItem()->node()).y() == 0
+					 && keyEvent->key() == Qt::Key_Up && hasTextField())
+				selectItem(textField());
+			//Otherwise just switch between the normal items
+			else if (focusedItem() != textField() || !hasTextField())
+			{
+				auto pos = indexOf(focusedItem()->node());
+				if (keyEvent->key() == Qt::Key_Down)
+					pos.setY(pos.y() + 1);
+				else if (keyEvent->key() == Qt::Key_Up)
+					pos.setY(pos.y() - 1);
+				else if (keyEvent->key() == Qt::Key_Left)
+					pos.setX(pos.x() - 1);
+				else
+					pos.setX(pos.x() + 1);
+				pos = correctCoordinates(pos);
+				selectItem(findVisualizationOf(currentNodes()[pos.x()][pos.y()]));
+			}
+			return true;
+		}
+		else if (keyEvent->key() == Qt::Key_Return)
+		{
+			event->accept();
+			if (executeFocused())
+			{
+				hideSelection();
+				return true;
+			}
+		}
+	}
+	else if (event->type() == QEvent::GraphicsSceneMousePress)
+	{
+		auto asItem = Visualization::Item::envisionItem(watched);
+		asItem = asItem ? asItem->findAncestorWithNode() : asItem;
+		if (asItem && isAncestorOf(watched))
+		{
+			selectItem(asItem);
```

Comment:
For example here, you might pass something to selectItem, which is not what you expect. If a menu entry contains a visualization that is based on node iteself, and I click on that visualization, I will get it in the `asItem` variable, whereas you expect the menu entry node.

One way to resolve this is to do it similarly to what `AutoCompleteVis` does, iterate over all entries and ask if any of them is the ancestor and if so execute it. It might take a few more cycles, but it's easier to understand in my opinion. Other suggestions are also welcome.